python/services/github_service.py

```diff
@@ -670,3 +698,184 @@ def remove_user_from_gitub(self, user: str):
         github_user = self.github_client_core_api.get_user(user)
         self.github_client_core_api.get_organization(
             self.organisation_name).remove_from_membership(github_user)
+
+    def report_on_inactive_users(self, teams: dict[str, dict[str, Any]], inactivity_months: int, slack_token: str) -> None:
+        """
+        Processes a list of GitHub teams to identify inactive users based on their commit activity within the given time frame.
+
+        For each team, the method:
+        - Loads the specific configuration related to that team (e.g., whether to remove users from the team, specific users or repositories to ignore, Slack channel for notifications, etc.).
+        - Retrieves the users and repositories associated with the team, considering any specified ignore settings.
+        - Determines which users are inactive according to the inactivity threshold in months.
+        - If configured to do so, removes inactive users from the team and/or accumulates them for later notification.
+        - If a Slack channel is configured for the team, sends a notification about the removed and/or to-be-removed users.
+
+        Parameters:
+        - teams (dict[str, dict[str, Any]]): A dictionary mapping team names to configuration dictionaries for processing each team.
+        - inactivity_months (int): The number of months of inactivity to use as the threshold for determining whether a user is considered inactive.
+        - slack_token (str): The authentication token for sending messages to Slack if configured for a team.
+
+        Returns:
+        None: This method performs its actions as side effects, updating team memberships and sending Slack notifications as configured.
+        """
+        users_to_remove = []
+        users_removed = []
+
+        for team_name, team_config in teams.items():
+            logging.info(f"Processing team {team_name}")
+            config = self._load_team_config(team_config)
+
+            users = self._get_users_from_team(config.github_team)
+            repositories = self._get_repositories_from_team(
+                config.github_team, config.ignore_repositories)
+
+            users_removed, users_to_remove = self._process_users(
+                users, repositories, config, inactivity_months)
+
+            if config.slack_channel and users_to_remove or config.slack_channel and users_removed:
+                logging.info(
+                    f"Sending message to slack channel {config.slack_channel}")
+                SlackService(slack_token).send_message_to_plaintext_channel_name(self._message_to_users(
+                    users_removed, users_to_remove, config.github_team, inactivity_months), config.slack_channel)
```

Comment:
I'd strongly recommend removing Slack references from the GithubService, ideally, the GithubService is only aware of Github as it's purpose is to 'orchestrate conversations with Github'

A few benefits of doing this:
- Easier to test
- Keeps GithubService responsibility well defined
- Keeps functions composable 
  - enabling us to re-use functionality and solve more complex problems easier
  - enabling us to easily add additional functionality i.e. adding GovUK notify notifications will be easier
- Side effect free 
- Reduces number of required to execute function
- Helps raise potential conditional logic to higher levels, i.e adding in a flag in the job to enable/disable slack notifications would exist at the top level instead of the Service layer

I'd be happy to dive into more detail here!